master/internal/provisioner/aws_spot.go

```diff
@@ -0,0 +1,683 @@
+
+	for _, instanceId := range instanceIDs {
+		if strings.HasPrefix(*instanceId, spotRequestIdPrefix) {
+			spotRequestId := instanceId
+			pendingSpotReqsToTerminate.add(*spotRequestId)
+		} else {
+			instancesToTerminate.add(*instanceId)
+		}
+	}
+
+	ctx.Log().
+		WithField("log-type", "terminateSpot.start").
+		Debugf(
+			"terminating %d EC2 instances and %d spot requests: %s,  %s",
+			instancesToTerminate.length(),
+			pendingSpotReqsToTerminate.length(),
+			instancesToTerminate.string(),
+			pendingSpotReqsToTerminate.string(),
+		)
+
+	terminateInstancesResponse, err := c.terminateInstances(instancesToTerminate.asListOfPointers(), false)
+	if err != nil {
+		ctx.Log().WithError(err).Error("cannot terminate EC2 instances")
+	} else {
+		terminated := c.newInstancesFromTerminateInstancesOutput(terminateInstancesResponse)
+		ctx.Log().
+			WithField("log-type", "terminateSpot.terminatedInstances").
+			Debugf(
+				"terminated %d EC2 instances: %s",
+				len(terminated),
+				fmtInstances(terminated),
+			)
+	}
+
+	_, err = c.terminateSpotInstanceRequests(ctx, pendingSpotReqsToTerminate.asListOfPointers(), false)
+	if err != nil {
+		ctx.Log().WithError(err).Error("cannot terminate spot requests")
+	} else {
+		ctx.Log().
+			WithField("log-type", "terminateSpot.terminatedSpotRequests").
+			Debugf(
+				"terminated %d spot requests: %s",
+				pendingSpotReqsToTerminate.length(),
+				pendingSpotReqsToTerminate.string(),
+			)
+	}
+
+	// TODO: Race condition - an instance could have been created between listing
+	//       and terminating spot request. We could clean up the spot instances here.
+	//       But it will be cleaned up on the next call to list() anyway.
+}
+
+func (c *awsCluster) launchSpot(
+	ctx *actor.Context,
+	instanceType instanceType,
+	instanceNum int,
+) {
+	instType, ok := instanceType.(ec2InstanceType)
+	if !ok {
+		panic("cannot pass non-ec2InstanceType to ec2Cluster")
+	}
+
+	if instanceNum < 0 {
+		return
+	}
+
+	ctx.Log().
+		WithField("log-type", "launchSpot.start").
+		Infof("launching %d EC2 spot requests", instanceNum)
+	resp, err := c.createSpotInstanceRequestsCorrectingForClockSkew(ctx, instanceNum, false, instType)
+	if err != nil {
+		ctx.Log().WithError(err).Error("cannot launch EC2 spot requests")
+		return
+	}
+
+	// Update the internal spotRequest tracker because there can be a large delay
+	// before the API starts including these requests in listSpotRequest API calls,
+	// and if we don't track it internally, we will end up overprovisioning.
+	for _, request := range resp.SpotInstanceRequests {
+		c.spot.trackedReqs.add(&spotRequest{
+			SpotRequestId: *request.SpotInstanceRequestId,
+			State:         *request.State,
+			StatusCode:    request.Status.Code,
+			StatusMessage: request.Status.Message,
+			CreationTime:  *request.CreateTime,
+			InstanceId:    nil,
+		})
+
+		ctx.Log().
+			WithField("log-type", "launchSpot.creatingRequest").
+			Infof(
+				"Creating spot request, %s, %s",
```

Comment:
can you change it to be "creating spot request: %s (state %s)" for consistency with other loggings?

Reply:
Done